libc/signal.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Phoenix-RTOS
+ *
+ * POSIX.1-2017 standard library functions tests
+ * HEADER:
+ *    - signal.h
+ *
+ * TESTED:
+ *    - psignal()
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Mateusz Bloch
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <signal.h>
+#include <limits.h>
+#include <unistd.h>
+#include <unity_fixture.h>
+
+/* Typical error message does not exceed ~60 characters, that's why we expect a maximum value a little bit bigger */
+#define MAX_LEN_STRING 100
+#define STDERR_PATH    "stderr.log"
+#define SIMPLE_MSG     "Simple Message"
+
+static FILE *err_file;
+static int stderr_fd;
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-const-variable"
+static const int signal_codes[] = { SIGABRT, SIGALRM, SIGBUS, SIGCHLD, SIGCONT, SIGFPE, SIGHUP, SIGILL, SIGINT,
+	SIGKILL, SIGPIPE, SIGQUIT, SIGSEGV, SIGSTOP, SIGTERM, SIGTSTP, SIGTTIN, SIGTTOU, SIGUSR1, SIGUSR2, SIGPROF,
+	SIGSYS, SIGTRAP, SIGURG, SIGVTALRM, SIGXCPU, SIGXFSZ };
+
+static const unsigned int signal_codes_len = sizeof(signal_codes) / sizeof(signal_codes[0]);
+#pragma GCC diagnostic pop
+
+TEST_GROUP(signal_psignal);
+
+TEST_SETUP(signal_psignal)
+{
+	err_file = fopen(STDERR_PATH, "a+");
+	stderr_fd = dup(STDERR_FILENO);
+}
+
+
+TEST_TEAR_DOWN(signal_psignal)
+{
+	TEST_ASSERT_NOT_NULL(freopen("/dev/stderr", "w", stderr));
+	close(STDERR_FILENO);
+	dup2(stderr_fd, STDERR_FILENO);
+	close(stderr_fd);
+	fclose(err_file);
+	remove(STDERR_PATH);
+}
+
+
+TEST(signal_psignal, basic)
+{
+/* Disabled because of #695 issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/695 */
+#ifdef __phoenix__
+	TEST_IGNORE_MESSAGE("#695 issue");
+#else
+	/* Since these function shall not return a value, we check only errno */
+	char buff[MAX_LEN_STRING];
+
+	for (int i = 0; i < signal_codes_len; i++) {
+		rewind(err_file);
+		memset(buff, 0, sizeof(buff));
+
+		TEST_ASSERT_NOT_NULL(freopen(STDERR_PATH, "w", stderr));
+
+		errno = 0;
+		psignal(signal_codes[i], "Lorem Ipsum");
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		fflush(stderr);
+
+		TEST_ASSERT_NOT_NULL(fgets(buff, MAX_LEN_STRING, err_file));
+		TEST_ASSERT_NOT_NULL(strstr(buff, "Lorem Ipsum"));
+		fflush(err_file);
+	}
+#endif
+}
+
+
+TEST(signal_psignal, ascii_string)
+{
+/* Disabled because of #695 issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/695 */
+#ifdef __phoenix__
+	TEST_IGNORE_MESSAGE("#695 issue");
+#else
+	char buff[MAX_LEN_STRING];
+
+	char ascii_str[MAX_LEN_STRING];
+
+	/*
+	 * In ASCII table from number 33 starts printable characters
+	 * in this case we want to dodge terminating ASCII signs
+	 */
+	for (int i = 33; i < 128; i++) {
```

Comment:
better, but `'0' + i` is still a bad idea. Also - try avoiding repeated magic numbers.

Maybe:
```c
char ascii_str[MAX_LEN_STRING] = {0};

/* use safely-printable chars */
const char first_char = '!';
for (char i = first_char; i < 120; i++) {
  ascii_str[i - first_char] = i;
}
// note: \0 is taken care of because of zeroing the array beforehand
```

OR:
```c
char c = 0, count = 0;
while (c < 128) {
  if (isprint(c)) {
   ascii_str[count++] = c;
 }
 c += 1;
}

ascii_str[count++] = '\0';
```